File: drivers/java/src/org/conf2/schema/browse/Walk.java
package org.conf2.schema.browse;

import org.conf2.schema.*;
import org.conf2.schema.driver.Driver;
import org.conf2.schema.driver.DriverHandle;

import java.util.Collections;
import java.util.Iterator;

/**
 *
 */
public class Walk {
//    private Driver driver;
//    public void insert(Selection from, Selection to) {
//        DriverHandle dfrom = getSelectionHandle(from);
//        DriverHandle dto = getSelectionHandle(from);
//
//
//    }
//
//
//    DriverHandle getSelectionHandle(Browser b, String path, Selection s) {
//
//
//    }

    public static Selection walk(Selection selection, BrowsePath path) {
        PathWalkController controller = new PathWalkController(path);
        walk(selection, controller);
        return controller.target;
    }

    public static void walk(Selection selection, WalkController controller) {
        Walk.walk(selection, controller, 0);
    }

    static void walk(Selection s, WalkController c, int level) {
        if (s.meta instanceof MetaList && ! s.insideList) {
            MetaList l = (MetaList) s.meta;
            boolean hasMore = c.selectionIterator(s, level, true);
            while (hasMore) {
                s.insideList = true;
                walk(s, c, level);
                hasMore = c.selectionIterator(s, level, false);
            }
        } else {
            Selection child;
            Iterator<Meta> i = c.containerIterator(s, level);
            while (i.hasNext()) {
                s.position = i.next();

                // TODO: resolve choice

                if (MetaUtil.isLeaf(s.position)) {
                    BrowseValue v = new BrowseValue();
                    // just walking, not interested in value
                    s.Read.Read();
                } else {
                    child = s.Enter.Enter();
                    if (!s.found) {
                        continue;
                    }
                    child.meta = (MetaCollection) s.position;
                    walk(child, c, level + 1);

                    if (s.Exit != null) {
                        s.Exit.Exit();
                    }
                }
            }
        }
    }
}

File: src/schema/browse/walk.go
package browse

import (
	"schema"
)

func WalkPath(selection *Selection, path *Path) (*Selection, error) {
	finder := NewFindTarget(path)
	err := walk(selection, finder)
	return finder.Target, err
}

func Walk(selection *Selection, controller WalkController) (err error) {
	return walk(selection, controller)
}

func walk(selection *Selection, controller WalkController) (err error) {
	state := selection.State
	if schema.IsList(state.SelectedMeta()) && !state.InsideList() {
		var next *Selection
		if next, err = controller.ListIterator(selection, true); err != nil {
			return
		}
		for next != nil {
			if err = walk(next, controller); err != nil {
				return
			}
			if err = selection.Fire(NEXT); err != nil {
				return err
			}
			if next, err = controller.ListIterator(selection, false); err != nil {
				return
			}
		}
	} else {
		var child Node
		i := controller.ContainerIterator(selection)
		for i.HasNextMeta() {
			state.SetPosition(i.NextMeta())
			if choice, isChoice := state.Position().(*schema.Choice); isChoice {
				var chosen schema.Meta
				if chosen, err = selection.Node.Choose(selection, choice); err != nil {
					return
				}
				state.SetPosition(chosen)
			}
			if schema.IsLeaf(state.Position()) {
				// only walking here, not interested in value
				if _, err = selection.Node.Read(selection, state.Position().(schema.HasDataType)); err != nil {
					return err
				}
			} else {
				metaList := state.Position().(schema.MetaList)
				if schema.IsAction(state.Position()) {
					if err = controller.VisitAction(selection); err != nil {
						return err
					}
				} else {
					if child, err = selection.Node.Select(selection, metaList, false); err != nil {
						return err
					}
					if child == nil {
						continue
					}
					defer schema.CloseResource(child)
					childSel := selection.Select(child)
					if err = walk(childSel, controller); err != nil {
						return
					}
					if err = childSel.Fire(LEAVE); err != nil {
						return err
					}
				}
			}
		}
	}
	return
}
